### ISSUE TITLE
[java] CouplingBetweenObjectsRule does not count attribute and local variable couplings properly

### ISSUE THREAD

**Issue by gustavokrieger** [pmd/pmd#2818]
**Affects PMD Version:** 6.28.0

**Rule:** [CouplingBetweenObjects](https://pmd.github.io/latest/pmd_rules_java_design.html#couplingbetweenobjects)

**Description:** The rule returns false negative in cases when the coupling is in an attribute or in a local variable, apparently counting multiple occurrences of these types of coupling as one. It works perfectly when the coupling is in the return type. From what I understood, [CouplingBetweenObjects.xml](https://github.com/pmd/pmd/blob/master/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/design/xml/CouplingBetweenObjects.xml) only tests with "expected-problems" in a code with couplings in the return type, possibly explaining why the issue went unnoticed.

**Code Sample demonstrating the issue:**

```java
import java.util.ArrayList;
import java.util.Vector;
import org.A;
import org.B;

public class Foo {

  public static final ArrayList arrayList = new ArrayList();
  public static final Vector vector = new Vector();

  public void doWork() {
    A a = new A();
    B b = new B();
  }

}
```

**Expected outcome:**

*   With this code as input, I believe [CouplingBetweenObjectsRule.java](https://github.com/pmd/pmd/blob/d5729e65dc838bf963a85b4aae61cb87e44b811e/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/CouplingBetweenObjectsRule.java) should count 4 coupling occurrences (ArrayList, Vector, A and B), but it counts 1 occurrence. False negative.

**Running PMD through:** *Maven*

**Comment by gustavopcassol** [comment]
I see that there is a check for repeated types before "couplingCount" is updated:

https://github.com/pmd/pmd/blob/47d3e1b9f98362d968a8dba7ecb82b62a9edf1f4/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/CouplingBetweenObjectsRule.java#L149-L150

Shouldn't coupling be type independent? Shouldn't "couplingCount" be updated every time a new attribute, local variable or return type is declared, and it's type is a class (as opposed to also counting primitives, void, etc.)? If that is not the case, wouldn't the result always be the same as [ExcessiveImports](https://pmd.github.io/latest/pmd_rules_java_design.html#excessiveimports)?

**Comment by gustavopcassol** [comment]
I also found out that the parameter "variableType", when using this issue's code sample as input, is always "null" in the following method:

https://github.com/pmd/pmd/blob/47d3e1b9f98362d968a8dba7ecb82b62a9edf1f4/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/CouplingBetweenObjectsRule.java#L133-L153

Which is why the outcome counts only 1 occurrence. The "null" in "variableType" is added to "typesFoundSoFar" in the following line:

https://github.com/pmd/pmd/blob/47d3e1b9f98362d968a8dba7ecb82b62a9edf1f4/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/CouplingBetweenObjectsRule.java#L151

And the subsequent occurrences (which also have "null" in their "variableType") can't get past the condition in the following line:

https://github.com/pmd/pmd/blob/47d3e1b9f98362d968a8dba7ecb82b62a9edf1f4/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/CouplingBetweenObjectsRule.java#L149

**Comment by jsotuyod** [comment]
This is no longer reproducible under PMD 7.0.0
